### PR TITLE
Add aten mkldnn ops: relu, max_pool2d and avg_pool2d

### PR DESCRIPTION
**Pull request by bddppq**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19633 Add is_mkldnn to at::Tensor&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D15053320/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19204 Add aten mkldnn conv2d operator&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14857513/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#19205 Add aten mkldnn ops: relu, max_pool2d and avg_pool2d**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14850598/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19206 Add aten mkldnn batch_norm operator&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14887205/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19207 Add aten mkldnn add operator&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14889477/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19209 Add aten mkldnn view operator&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14894545/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19210 Add aten mkldnn linear operator&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14901641/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19648 Adjust resnext run script&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14889481/)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/19205

Differential Revision: [D14850598](https://our.internmc.facebook.com/intern/diff/D14850598/)

### CODE REVIEW

File: aten/src/ATen/native/mkldnn/Pooling.cpp [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+      input,
+      kernel_size,
+      stride,
+      padding,
+      dilation,
+      ceil_mode,
+      ideep::algorithm::pooling_max);
+  return std::make_tuple(
+      output,
+      // TODO: ideep currently doesn't expose the max indices,
```
Comment (gchanan) [inline aten/src/ATen/native/mkldnn/Pooling.cpp]: umm...would it work to just throw an exception if they ask for the indices?

**Review by gchanan — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -3705,10 +3721,18 @@
...
+  dispatch:
+    CPU: max_pool2d_with_indices_out
+    CUDA: max_pool2d_with_indices_out
+    MkldnnCPU: mkldnn_max_pool2d_with_indices_out
 
 # Return: (Tensor output, Tensor indices)
 - func: max_pool2d_with_indices(Tensor self, int[2] kernel_size, int[2] stride=[], int[2] padding=0, int[2] dilation=1, bool ceil_mode=False) -> (Tensor, Tensor)
   python_module: nn
+  dispatch:
+    CPU: max_pool2d_with_indices
```
[the remark this answers is not in the document]
Comment (bddppq) [reply aten/src/ATen/native/native_functions.yaml]: nope because the derivative of max_pool2d_with_indices has been defined in `tools/autograd/derivatives.yaml`

File: torch/nn/modules/activation.py [hunk 3]
```diff
@@ -92,6 +92,10 @@ class ReLU(Threshold):
     def __init__(self, inplace=False):
         super(ReLU, self).__init__(0., 0., inplace)
 
+    @weak_script_method
+    def forward(self, input):
+        return F.relu(input, inplace=self.inplace)
```
[the remark this answers is not in the document]
Comment (bddppq) [reply torch/nn/modules/activation.py]: idk :-) will see from the CI. the main reason I change it to directly call relu is because threshold's semantics is too hard to implement in mkldnn.

File: aten/src/ATen/native/mkldnn/Pooling.cpp (hunk 1 above)
Comment (bddppq) [reply aten/src/ATen/native/mkldnn/Pooling.cpp]: There is no way to know whether indices is required or not in c++. We *could* throw in python side (where `return_indices` is an argument) but that would require exposing is_mkldnn to python side (not necessary a bad idea).
We *could* get the indices from mkldnn as well from a non-public api (@Jianhui-Li mentioned we can get it from the ideep::tensor `get_extra`)

Comment (jgong5) [reply aten/src/ATen/native/mkldnn/Pooling.cpp]: Yes, we can use `get_extra` of the output ideep tensor. See this:
https://github.com/intel/ideep/blob/960ee3a20d975655d3fa8e40e445898a1e3b4078/include/ideep/computations.hpp#L3882

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -3711,10 +3727,18 @@
 # Return: (Tensor output, Tensor indices)
 - func: max_pool2d_with_indices(Tensor self, int[2] kernel_size, int[2] stride=[], int[2] padding=0, int[2] dilation=1, bool ceil_mode=False, *, Tensor(a!) output, Tensor(b!) indices) -> (Tensor(a!), Tensor(b!))
   python_module: nn
+  dispatch:
+    CPU: max_pool2d_with_indices_out
+    CUDA: max_pool2d_with_indices_out
+    MkldnnCPU: mkldnn_max_pool2d_with_indices_out
 
 # Return: (Tensor output, Tensor indices)
 - func: max_pool2d_with_indices(Tensor self, int[2] kernel_size, int[2] stride=[], int[2] padding=0, int[2] dilation=1, bool ceil_mode=False) -> (Tensor, Tensor)
```
Comment (jgong5) [inline aten/src/ATen/native/native_functions.yaml]: BTW: There is also `max_pool2d` that doesn't need to return indices, useful for inference workload. We can implement it more efficiently with MKL-DNN by setting the `prop_kind` to `forward_inference`.

File: aten/src/ATen/native/mkldnn/Pooling.cpp (hunk 1 above)
Comment (mingfeima) [reply aten/src/ATen/native/mkldnn/Pooling.cpp]: 1. the problem here is that `indice` of mkldnn pooling is different from what PyTorch looks like, both the **dtype** and **content**.
a. mkldnn - `indice` is a local view w.r.t. kernel, the `dtype` could be either `int32_t` or `int8_t`
b. PyTorch - `indice` is a global view w.r.t. input.
See [mkldnn_issue_259](https://github.com/intel/mkl-dnn/issues/259) for detail.
So directly returning mkldnn indice is meaningless. The mkldnn indice is more like a cache used by backward during training. Well in that case, it should be handled by `workspace` in ATen.
2. at the current moment, probably the most simple way to address this is remove mkldnn path for `mkldnn_max_pool2d_with_indices_out`. Which means, only when `indice` is not in the output, mkldnn is used.
This won't make any difference for models like `ResNet` or `ResNext`.
3. in case we need to support unpooling (up-sampling) where `indice` is needed, we can convert the mkldnn `indice` to global by doing the math pointed in the link above.

File: aten/src/ATen/native/mkldnn/Utils.cpp [hunk 5]
```diff
@@ -28,4 +29,30 @@ std::vector<int64_t> conv_output_size(
...
+    output_size[i] = pooling_output_shape<int64_t>(
+      input_size[i],
+      kernel_size[i - 2],
+      padding[i - 2],
+      stride[i - 2],
+      dilation[i - 2],
+      ceil_mode
+    );
+  }
+
```
Comment (mingfeima) [inline aten/src/ATen/native/mkldnn/Utils.cpp]: MKLDNN treats padding differently with ceil mode.
e.g. the following case would result runtime error, both max pooling and avg pooling:
```python
import torch
x = torch.randn(1, 1, 10, 10, dtype=torch.float32) * 10
max_pool2d = torch.nn.MaxPool2d(kernel_size=(4,5),stride=(3,4),padding=(1,2), ceil_mode=True)
# max_pool2d = torch.nn.AvgPool2d(kernel_size=(4,5),stride=(3,4),padding=(1,2), ceil_mode=True)
output =  max_pool2d(x.to_mkldnn()).to_dense()
```
please refer [caffe ](https://github.com/intel/caffe/blob/master/src/caffe/layers/mkldnn_pooling_layer.cpp#L122) code of computing pooling output size for detail. @XiaobingSuper

File: aten/src/ATen/native/mkldnn/Pooling.cpp (hunk 1 above)
Comment (jgong5) [reply aten/src/ATen/native/mkldnn/Pooling.cpp]: @bddppq Due to the differences between PyTorch indices tensor (using global indices) and MKL-DNN indices tensor (using local indices), we are considering to hide the MKL-DNN indices inside ideep as an "opaque" format. Users can get PyTorch indices with `to_dense()`. This may take a while and before that happens, how about we only implement `max_pool2d` without indices in MKL-DNN path as @mingfeima proposed?

File: torch/nn/modules/activation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/nn/modules/activation.py]: Why do we still inherit from `Threshold` if we replace `forward` anyway?

File: aten/src/ATen/native/mkldnn/Pooling.cpp (hunk 1 above)
Comment (dzhulgakov) [reply aten/src/ATen/native/mkldnn/Pooling.cpp]: @bddppq - I think everyone would be calling max pool from nn.functional, not as the direct function in torch._C._nn. So I think we can change https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L489 to just call torch._C._nn.max_pool_2d and overload that function to go to mkldnn implementation. Would that work?

Comment (bddppq) [reply aten/src/ATen/native/mkldnn/Pooling.cpp]: aten max_pool2d right now doesn’t have dispatch (and is not even exposed to python), so in order to do that we will need to add derivatives for it to let codegen do the unwrap thing for it.

Comment (dzhulgakov) [reply aten/src/ATen/native/mkldnn/Pooling.cpp]: If you just want unwrap - there's 'require_tensor: True' annotation you can put in native_functions.yaml

As for exposing to python, I think if you make it as 'python_module: nn' or somethingl ike this, it'd be exposed only as torch._C._nn - I think it's ok to have it but not call it from functional wrappers

Comment (bddppq) [reply aten/src/ATen/native/mkldnn/Pooling.cpp]: It's explicitly not exposed to python: https://github.com/pytorch/pytorch/blob/1c5073f/tools/autograd/gen_python_functions.py#L33
I don't dare to open this can of worms (actually I did at the very beginning and then ran into a test failure complaining ```derivative for max_pool2d is not implemented``` (or similar)).

Since
1. indices are only used in training
2. mkldnn/ideep will add support to expose indices in their api

Let's just return some garbage values and only support inference for now.

File: aten/src/ATen/native/mkldnn/Utils.cpp (hunk 5 above)
Comment (dzhulgakov) [reply aten/src/ATen/native/mkldnn/Utils.cpp]: has this comment been addressed? (sorry I didn't follow the point)

File: aten/src/ATen/native/mkldnn/Pooling.cpp (hunk 1 above)
Comment (dzhulgakov) [reply aten/src/ATen/native/mkldnn/Pooling.cpp]: Hm, I think it should have worked if we were to check requires_grad. But I don't insist.

Can you please expand this comment with more details on why garbage is fine? (pretty much what we have discussed above)

**Review by dzhulgakov — APPROVED** [review]
Looks good as long as the sizes calculation comment has been addressed

File: aten/src/ATen/native/mkldnn/Pooling.cpp (hunk 1 above)
Comment (bddppq) [reply aten/src/ATen/native/mkldnn/Pooling.cpp]: Actually @wanchaol is working on exposing max_pool2d to python https://github.com/pytorch/pytorch/pull/19449. I will give a try to implement mkldnn for max_pool2d once that is merged

File: aten/src/ATen/native/mkldnn/Utils.cpp (hunk 5 above)
Comment (XiaobingSuper) [reply aten/src/ATen/native/mkldnn/Utils.cpp]: This problem can be solved by change the mkldnn padding as caffe's doing: https://github.com/intel/caffe/blob/362a3b375e8f5bc64f5338d18f1742448cfb8fb7/src/caffe/layers/mkldnn_pooling_layer.cpp#L148

Comment (XiaobingSuper) [reply aten/src/ATen/native/mkldnn/Utils.cpp]: Please note that mkldnn is to get given output size by changing padding size.

Comment (bddppq) [reply aten/src/ATen/native/mkldnn/Utils.cpp]: Added an assert to only support non ceil_mode for now. Let's do the padding trick in a follow up PR.

File: aten/src/ATen/native/mkldnn/MKLDNNCommon.cpp [hunk 6]
```diff
@@ -43,7 +43,6 @@ using MKLDNNTensor = Tensor;
 Tensor new_with_itensor_mkldnn(ideep::tensor&& it, const TensorOptions& options) {
   // NOTE: int32_t dims from ideep::tensor but sizes needs int64_t
   // TODO: support int64_t dims in ideep::tensor to avoid extra conversion
-  AT_ASSERT(!it.has_extra());
```
Comment (dzhulgakov) [inline aten/src/ATen/native/mkldnn/MKLDNNCommon.cpp]: why did we originally have it here?

File: aten/src/ATen/native/mkldnn/Pooling.cpp [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+Tensor& mkldnn_avg_pool2d_out(
+    Tensor& output,
+    const Tensor& input,
+    IntArrayRef kernel_size,
+    IntArrayRef stride,
+    IntArrayRef padding,
+    bool ceil_mode,
+    bool count_include_pad) {
+  output = _mkldnn_pool2d(
```
Comment (dzhulgakov) [inline aten/src/ATen/native/mkldnn/Pooling.cpp]: hm, I think it's wrong. _out variants are supposed to assign things in-place in the same Tensor instance, no? @gchanan to double-check

and I'd add a test for it too

File: aten/src/ATen/native/native_functions.yaml [hunk 8]
```diff
@@ -1136,6 +1136,11 @@
 
 - func: max_pool2d(Tensor self, int[2] kernel_size, int[2] stride=[], int[2] padding=0, int[2] dilation=1, bool ceil_mode=False) -> Tensor
 
+- func: mkldnn_max_pool2d(Tensor self, int[2] kernel_size, int[2] stride=[], int[2] padding=0, int[2] dilation=1, bool ceil_mode=False) -> Tensor
+  requires_tensor: True
```
Comment (dzhulgakov) [inline aten/src/ATen/native/native_functions.yaml]: requires_tensor is probably not required here

File: aten/src/ATen/native/mkldnn/Pooling.cpp (hunk 7 above)
Comment (bddppq) [reply aten/src/ATen/native/mkldnn/Pooling.cpp]: I will add a test case, but IIRC the _out version doesn't require same instance (the "_" version does).

File: aten/src/ATen/native/native_functions.yaml (hunk 8 above)
Comment (bddppq) [reply aten/src/ATen/native/native_functions.yaml]: yep just for documentation purpose

File: aten/src/ATen/native/mkldnn/Pooling.cpp (hunk 7 above)
Comment (bddppq) [reply aten/src/ATen/native/mkldnn/Pooling.cpp]: You are right _out requires same identity. We will need to change opaque tensor impl to allow setting sizes to support this. In this PR let's not support in-place avg_pool for now.

Comment (dzhulgakov) [reply aten/src/ATen/native/mkldnn/Pooling.cpp]: Usually _out versions just fail if the size is wrong. But I agree that we can just skip _out version here

Comment (bddppq) [reply aten/src/ATen/native/mkldnn/Pooling.cpp]: Oh so normally the output tensor got passed in is supposed to already have the right sizes? (and other meta info like dtype)
